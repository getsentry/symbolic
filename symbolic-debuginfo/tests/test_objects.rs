use std::{ffi::CString, fmt};

use symbolic_common::ByteView;
use symbolic_debuginfo::{elf::ElfObject, FileEntry, Function, Object, SymbolMap};
use symbolic_testutils::fixture;

use similar_asserts::assert_eq;

type Error = Box<dyn std::error::Error>;

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for file lists.
struct FilesDebug<'a>(&'a [FileEntry<'a>]);

impl fmt::Debug for FilesDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for file in self.0 {
            writeln!(f, "{}", file.abs_path_str())?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open(fixture("windows/crash.sym"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Breakpad(
   ⋮    BreakpadObject {
   ⋮        code_id: Some(
   ⋮            CodeId(5ab380779000),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
   ⋮            appendix: 1,
   ⋮        },
   ⋮        arch: X86,
   ⋮        name: "crash.pdb",
   ⋮        has_symbols: true,
   ⋮        has_debug_info: true,
   ⋮        has_unwind_info: true,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.sym"))?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_files() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.sym"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let files = session.files().collect::<Result<Vec<_>, _>>()?;
    assert_eq!(files.len(), 147);
    insta::assert_debug_snapshot!("breakpad_files", FilesDebug(&files[..10]));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.sym"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot!("breakpad_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_breakpad_functions_mac_with_inlines() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash.inlines.sym"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot!(
        "breakpad_functions_mac_with_inlines",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/crash"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Elf(
   ⋮    ElfObject {
   ⋮        code_id: Some(
   ⋮            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
   ⋮            appendix: 0,
   ⋮        },
   ⋮        arch: Amd64,
   ⋮        kind: Executable,
   ⋮        load_address: 0x400000,
   ⋮        has_symbols: true,
   ⋮        has_debug_info: false,
   ⋮        has_unwind_info: true,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/crash.debug"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Elf(
   ⋮    ElfObject {
   ⋮        code_id: Some(
   ⋮            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
   ⋮            appendix: 0,
   ⋮        },
   ⋮        arch: Amd64,
   ⋮        kind: Debug,
   ⋮        load_address: 0x400000,
   ⋮        has_symbols: true,
   ⋮        has_debug_info: true,
   ⋮        has_unwind_info: false,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/crash.debug"))?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_files() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/crash.debug"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let files = session.files().collect::<Result<Vec<_>, _>>()?;
    assert_eq!(files.len(), 1012);
    insta::assert_debug_snapshot!("elf_files", FilesDebug(&files[..10]));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/crash.debug"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

fn elf_debug_crc() -> Result<u32, Error> {
    Ok(u32::from_str_radix(
        std::fs::read_to_string(fixture("linux/elf_debuglink/gen/debug_info.txt.crc"))?.trim(),
        16,
    )?)
}

fn check_debug_info(filename: &'static str, debug_info: &'static str) -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/elf_debuglink/gen/".to_owned() + filename))?;

    let object = ElfObject::parse(&view)?;

    let debug_link = object
        .debug_link()
        .map_err(|err| err.kind)?
        .expect("debug link not found");

    assert_eq!(debug_link.filename(), CString::new(debug_info)?.as_c_str(),);
    assert_eq!(debug_link.crc(), elf_debug_crc()?);

    Ok(())
}

#[test]
fn test_elf_debug_link() -> Result<(), Error> {
    check_debug_info("elf_with_debuglink", "debug_info.txt")
}

#[test]
fn test_elf_debug_link_none() -> Result<(), Error> {
    let view = ByteView::open(fixture("linux/elf_debuglink/gen/elf_without_debuglink"))?;

    let object = ElfObject::parse(&view)?;

    let debug_link = object.debug_link().map_err(|err| err.kind)?;
    assert!(debug_link.is_none(), "debug link unexpectedly found");

    Ok(())
}

#[test]
// Test that the size of the debug_info filename doesn't influence the result.
fn test_elf_debug_link_padding() -> Result<(), Error> {
    check_debug_info("elf_with1_debuglink", "debug_info1.txt")?;
    check_debug_info("elf_with12_debuglink", "debug_info12.txt")?;
    check_debug_info("elf_with123_debuglink", "debug_info123.txt")
}

#[test]
// Test with a compressed gnu_debuglink section. This exerts the "Owned" path of the code,
// while without compression the "Borrowed" is exerted.
fn test_elf_debug_link_compressed() -> Result<(), Error> {
    check_debug_info("elf_with_compressed_debuglink", "debug_info.txt")
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮MachO(
   ⋮    MachObject {
   ⋮        code_id: Some(
   ⋮            CodeId(67e9247c814e392ba027dbde6748fcbf),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
   ⋮            appendix: 0,
   ⋮        },
   ⋮        arch: Amd64,
   ⋮        kind: Executable,
   ⋮        load_address: 0x100000000,
   ⋮        has_symbols: true,
   ⋮        has_debug_info: false,
   ⋮        has_unwind_info: true,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash.dSYM/Contents/Resources/DWARF/crash"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮MachO(
   ⋮    MachObject {
   ⋮        code_id: Some(
   ⋮            CodeId(67e9247c814e392ba027dbde6748fcbf),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
   ⋮            appendix: 0,
   ⋮        },
   ⋮        arch: Amd64,
   ⋮        kind: Debug,
   ⋮        load_address: 0x100000000,
   ⋮        has_symbols: true,
   ⋮        has_debug_info: true,
   ⋮        has_unwind_info: false,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash"))?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_files() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash.dSYM/Contents/Resources/DWARF/crash"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let files = session.files().collect::<Result<Vec<_>, _>>()?;
    assert_eq!(files.len(), 554);
    insta::assert_debug_snapshot!("mach_files", FilesDebug(&files[..10]));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view = ByteView::open(fixture("macos/crash.dSYM/Contents/Resources/DWARF/crash"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.exe"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Pe(
   ⋮    PeObject {
   ⋮        code_id: Some(
   ⋮            CodeId(5ab380779000),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
   ⋮            appendix: 1,
   ⋮        },
   ⋮        debug_file_name: Some(
   ⋮            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
   ⋮        ),
   ⋮        arch: X86,
   ⋮        kind: Executable,
   ⋮        load_address: 0x400000,
   ⋮        has_symbols: false,
   ⋮        has_debug_info: false,
   ⋮        has_unwind_info: false,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/CrashWithException.exe"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Pe(
   ⋮    PeObject {
   ⋮        code_id: Some(
   ⋮            CodeId(5c9e09599000),
   ⋮        ),
   ⋮        debug_id: DebugId {
   ⋮            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
   ⋮            appendix: 1,
   ⋮        },
   ⋮        debug_file_name: Some(
   ⋮            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb",
   ⋮        ),
   ⋮        arch: Amd64,
   ⋮        kind: Executable,
   ⋮        load_address: 0x140000000,
   ⋮        has_symbols: false,
   ⋮        has_debug_info: false,
   ⋮        has_unwind_info: true,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot!(object, @r###"
   ⋮Pdb(
   ⋮    PdbObject {
   ⋮        debug_id: DebugId {
   ⋮            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
   ⋮            appendix: 1,
   ⋮        },
   ⋮        arch: X86,
   ⋮        load_address: 0x0,
   ⋮        has_symbols: true,
   ⋮        has_debug_info: true,
   ⋮        has_unwind_info: true,
   ⋮        is_malformed: false,
   ⋮    },
   ⋮)
    "###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_files() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let files = session.files().collect::<Result<Vec<_>, _>>()?;
    assert_eq!(files.len(), 967);
    insta::assert_debug_snapshot!("pdb_files", FilesDebug(&files[..10]));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pdb_anonymous_namespace() -> Result<(), Error> {
    // Regression test for ?A0x<hash> namespaces

    let view = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let main_function = session
        .functions()
        .filter_map(|f| f.ok())
        .find(|f| f.address == 0x2910)
        .expect("main function at 0x2910");

    let start_function = main_function
        .inlinees
        .iter()
        .find(|f| f.address == 0x2a3d)
        .expect("start function at 0x2a3d");

    // was: "?A0xc3a0617d::start"
    assert_eq!(start_function.name, "`anonymous namespace'::start");

    Ok(())
}

#[test]
fn test_wasm_symbols() -> Result<(), Error> {
    let view = ByteView::open(fixture("wasm/simple.wasm"))?;
    let object = Object::parse(&view)?;

    assert_eq!(
        object.debug_id(),
        "bda18fd8-5d4a-4eb8-9302-2d6bfad846b1".parse().unwrap()
    );
    assert_eq!(
        object.code_id(),
        Some("bda18fd85d4a4eb893022d6bfad846b1".into())
    );

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot!("wasm_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_wasm_line_program() -> Result<(), Error> {
    let view = ByteView::open(fixture("wasm/simple.wasm"))?;
    let object = Object::parse(&view)?;

    let session = object.debug_session()?;
    let main_function = session
        .functions()
        .filter_map(|f| f.ok())
        .find(|f| f.address == 0x8b)
        .expect("main function at 0x8b");

    assert_eq!(main_function.name, "internal_func");

    Ok(())
}
