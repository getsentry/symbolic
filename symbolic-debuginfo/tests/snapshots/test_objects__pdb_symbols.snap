---
source: symbolic-debuginfo/tests/test_objects.rs
expression: SymbolsDebug(&symbols)
---
            1000 ?RequestDump@CrashGenerationClient@google_breakpad@@QAE_NPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            1120 ??1?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAE@XZ
            1180 ?assign@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV12@QB_WI@Z
            12d0 ?_Xlen@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@SAXXZ
            12e0 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@$$QAV01@@Z
            1360 ___local_stdio_printf_options
            1370 _swprintf_s
            13b0 __snwprintf_s
            13f0 _fprintf
            1420 ?Initialize@ExceptionHandler@google_breakpad@@AAEXABV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@P6A_NPAXPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@ZP6A_NPB_W5123_N@Z1HW4_MINIDUMP_TYPE@@51PAVCrashGenerationClient@2@PBUCustomClientInfo@2@@Z
            1770 ??1ExceptionHandler@google_breakpad@@QAE@XZ
            1a80 ?ExceptionHandlerThreadMain@ExceptionHandler@google_breakpad@@CGKPAX@Z
            1b90 ??0AutoExceptionHandler@google_breakpad@@QAE@XZ
            1c00 ??1AutoExceptionHandler@google_breakpad@@QAE@XZ
            1c40 ?HandleException@ExceptionHandler@google_breakpad@@CGJPAU_EXCEPTION_POINTERS@@@Z
            1d60 ?HandleInvalidParameter@ExceptionHandler@google_breakpad@@CAXPB_W00II@Z
            1ef0 ?HandlePureVirtualCall@ExceptionHandler@google_breakpad@@CAXXZ
            2020 ?WriteMinidumpOnHandlerThread@ExceptionHandler@google_breakpad@@AAE_NPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            2100 ?WriteMinidumpWithException@ExceptionHandler@google_breakpad@@AAE_NKPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@@Z
            2180 ?MinidumpWriteDumpCallback@ExceptionHandler@google_breakpad@@CGHPAXQAU_MINIDUMP_CALLBACK_INPUT@@PAU_MINIDUMP_CALLBACK_OUTPUT@@@Z
            2210 ?WriteMinidumpWithExceptionForProcess@ExceptionHandler@google_breakpad@@AAE_NKPAU_EXCEPTION_POINTERS@@PAUMDRawAssertionInfo@@PAX_N@Z
            2440 ?UpdateNextID@ExceptionHandler@google_breakpad@@AAEXXZ
            2560 ??$emplace_back@PAVExceptionHandler@google_breakpad@@@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@QAEX$$QAPAVExceptionHandler@google_breakpad@@@Z
            2690 ??1?$scoped_ptr@VCrashGenerationClient@google_breakpad@@@google_breakpad@@QAE@XZ
            26a0 ??1?$list@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@QAE@XZ
            26f0 ??_GCrashGenerationClient@google_breakpad@@QAEPAXI@Z
            2780 ?_Xrange@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@CAXXZ
            2790 ?_Xlength@?$vector@PAVExceptionHandler@google_breakpad@@V?$allocator@PAVExceptionHandler@google_breakpad@@@std@@@std@@CAXXZ
            27a0 ?_Buynode0@?$_List_alloc@U?$_List_base_types@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@@std@@QAEPAU?$_List_node@UAppMemory@google_breakpad@@PAX@2@PAU32@0@Z
            27d0 ??$_Buynode@ABUAppMemory@google_breakpad@@@?$_List_buy@UAppMemory@google_breakpad@@V?$allocator@UAppMemory@google_breakpad@@@std@@@std@@QAEPAU?$_List_node@UAppMemory@google_breakpad@@PAX@1@PAU21@0ABUAppMemory@google_breakpad@@@Z
            27f0 ?GUIDToWString@GUIDString@google_breakpad@@SA?AV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@PAU_GUID@@@Z
            28a0 _printf
            2910 _main
            2a6e @__security_check_cookie@4
            2a7f ??2@YAPAXI@Z
            2aaf ??3@YAXPAXI@Z
            2abd ___raise_securityfailure
            2ae5 ___report_gsfailure
            2e15 _mainCRTStartup
            2e1f ??3@YAXPAX@Z
            2e24 ??0bad_alloc@std@@QAE@ABV01@@Z
            2e3f ??0bad_alloc@std@@QAE@XZ
            2e57 ??0bad_array_new_length@std@@QAE@ABV01@@Z
            2e72 ??0bad_array_new_length@std@@QAE@XZ
            2e8a ??0exception@std@@QAE@ABV01@@Z
            2eb6 ??1bad_alloc@std@@UAE@XZ
            2ec7 ??_Eexception@std@@UAEPAXI@Z
            2ef4 ?__scrt_throw_std_bad_alloc@@YAXXZ
            2f11 ?__scrt_throw_std_bad_array_new_length@@YAXXZ
            2f2e ?what@exception@std@@UBEPBDXZ
            2f7f ___scrt_acquire_startup_lock
            2fb4 ___scrt_initialize_crt
            2fed ___scrt_initialize_onexit_tables
            3097 ___scrt_is_nonwritable_in_current_image
            3121 ___scrt_release_startup_lock
            313e ___scrt_uninitialize_crt
            3166 __onexit
            31a1 _atexit
            3205 ___security_init_cookie
            3250 __matherr
            3253 __get_startup_argv_mode
            3257 __get_startup_file_mode
            325d ?__scrt_initialize_type_info@@YAXXZ
            3269 ___vcrt_uninitialize
            326c __initialize_default_precision
            328d @_guard_check_icall_nop@4
            328e ___local_stdio_scanf_options
            3294 ___scrt_initialize_default_local_stdio_options
            32b1 ___scrt_is_user_matherr_present
            32bd ___scrt_get_dyn_tls_init_callback
            32c3 ___scrt_get_dyn_tls_dtor_callback
            32c9 ___scrt_fastfail
            33e5 ___scrt_is_managed_app
            3429 ___scrt_set_unhandled_exception_filter
            3435 ___scrt_unhandled_exception_filter@4
            3476 __crt_debugger_hook
            347e __RTC_Initialize
            34a9 __RTC_Terminate
            34d4 @_guard_check_icall@4
            34e0 __SEH_prolog4
            3526 __SEH_epilog4
            353b __except_handler4
            355e ??_Gtype_info@@UAEPAXI@Z
            3581 ___isa_available_init
            371a ___scrt_is_ucrt_dll_in_use
            3726 ___CxxFrameHandler3
            372c ___std_exception_copy
            3732 ___std_exception_destroy
            3738 __CxxThrowException@8
            373e _memset
            3744 __except_handler4_common
            374a _exit
            3750 __callnewh
            3756 _malloc
            375c __seh_filter_exe
            3762 __set_app_type
            3768 ___setusermatherr
            376e __configure_narrow_argv
            3774 __initialize_narrow_environment
            377a __get_initial_narrow_environment
            3780 __initterm
            3786 __initterm_e
            378c __exit
            3792 __set_fmode
            3798 ___p___argc
            379e ___p___argv
            37a4 __cexit
            37aa __c_exit
            37b0 __register_thread_local_exe_atexit_callback
            37b6 __configthreadlocale
            37bc __set_new_mode
            37c2 ___p__commode
            37c8 _free
            37ce __initialize_onexit_table
            37d4 __register_onexit_function
            37da __crt_atexit
            37e0 __controlfp_s
            37e6 _terminate
            37ec _IsProcessorFeaturePresent@4
            37f2 _memcpy

